File: aat-lib/src/main/java/ch/bailu/aat_lib/map/layer/grid/ElevationLayer.java
package ch.bailu.aat_lib.map.layer.grid;


import org.mapsforge.core.model.LatLong;

import ch.bailu.aat_lib.service.InsideContext;
import ch.bailu.aat_lib.description.AltitudeDescription;
import ch.bailu.aat_lib.map.MapContext;
import ch.bailu.aat_lib.map.Point;
import ch.bailu.aat_lib.map.layer.MapLayerInterface;
import ch.bailu.aat_lib.preferences.StorageInterface;
import ch.bailu.aat_lib.service.ServicesInterface;

public final class ElevationLayer implements MapLayerInterface {

    private final static int MIN_ZOOM_LEVEL=7;

    private final AltitudeDescription altitudeDescription;
    private final ServicesInterface services;

    public ElevationLayer(ServicesInterface services, StorageInterface storage) {
        altitudeDescription= new AltitudeDescription(storage);
        this.services = services;
    }

    @Override
    public void onLayout(boolean changed, int l, int t, int r, int b) {}

    @Override
    public void drawInside(MapContext mcontext) {}

    @Override
    public void drawForeground(MapContext mcontext) {
        byte zoomLevel = (byte)mcontext.getMetrics().getZoomLevel();
        final LatLong point = mcontext.getMapView().getMapViewPosition().getCenter();

        drawElevation(services, mcontext, zoomLevel, point);
    }


    private void drawElevation(ServicesInterface sc, MapContext mc, int zoom, final LatLong point) {
        if (zoom > MIN_ZOOM_LEVEL) {
            new InsideContext(sc) {
                @Override
                public void run() {
                    final short ele = sc.
                            getElevationService().
                            getElevation(point.getLatitudeE6(), point.getLongitudeE6());
                    mc.draw().textBottom(altitudeDescription.getValueUnit(ele),2);

                }
            };
        }
    }

    @Override
    public boolean onTap(Point tapPos) {
        return false;
    }

    @Override
    public void onPreferencesChanged(StorageInterface s, String key) {}

    @Override
    public void onAttached() {}

    @Override
    public void onDetached() {}
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/map/layer/grid/ElevationLayer.kt
package ch.bailu.aat_lib.map.layer.grid

import ch.bailu.aat_lib.description.AltitudeDescription
import ch.bailu.aat_lib.map.MapContext
import ch.bailu.aat_lib.map.layer.MapLayerInterface
import ch.bailu.aat_lib.preferences.StorageInterface
import ch.bailu.aat_lib.service.ServicesInterface
import ch.bailu.aat_lib.util.Point
import org.mapsforge.core.model.LatLong

class ElevationLayer(private val services: ServicesInterface, storage: StorageInterface?) : MapLayerInterface {
    private val altitudeDescription: AltitudeDescription = AltitudeDescription(storage!!)

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {}
    override fun drawInside(mcontext: MapContext) {}
    override fun drawForeground(mcontext: MapContext) {
        val zoomLevel = mcontext.getMetrics().getZoomLevel().toByte()
        val point = mcontext.getMapView().getMapViewPosition().center
        drawElevation(services, mcontext, zoomLevel.toInt(), point)
    }

    private fun drawElevation(sc: ServicesInterface, mc: MapContext, zoom: Int, point: LatLong) {
        if (zoom > MIN_ZOOM_LEVEL) {
            sc.insideContext {
                val ele = sc.getElevationService().getElevation(point.latitudeE6, point.longitudeE6)
                mc.draw().textBottom(altitudeDescription.getValueUnit(ele.toFloat()), 2)
            }
        }
    }

    override fun onTap(tapPos: Point): Boolean {
        return false
    }

    override fun onPreferencesChanged(storage: StorageInterface, key: String) {}
    override fun onAttached() {}
    override fun onDetached() {}

    companion object {
        private const val MIN_ZOOM_LEVEL = 7
    }
}
